### PR TITLE
Improving class loading efficiency and considering dynamic type list

### PR DESCRIPTION
**Pull request by J007X**
This PR fixes https://github.com/asyml/forte/issues/908. 

### Possible influences of this PR.
By using a special non-cached version of get_class (get_class_nc) in "is_subclass" might cause a slight performance impact of a few percentage (to ensure correctness), in the future there might be improved ways to make the method using some kind of caching mechanism safely.

### Test Conducted
testPOSTaggingNER() method in "data_pack_profiling_test" for checking performance changes (performance improved from non-lru cached versions of "get_class" in utils.py)

### CODE REVIEW

**Comment by hunterhector** [comment]
please fix the "This PR fixes" line again. Please learn how to use this syntac

**Review by hunterhector — COMMENTED** [review]
Please fix the conflicts and fix the doc error

**Comment by J007X** [comment]
> please fix the "This PR fixes" line again. Please learn how to use this syntax

Sorry forgot to remove the "[ ]" in the template. The format is now fixed.

File: forte/utils/utils.py [hunk 1]
```diff
@@ -78,8 +79,10 @@ def get_class_name(o, lower: bool = False) -> str:
         return o.__name__
 
 
-def get_class(full_class_name: str, module_paths: Optional[List[str]] = None):
-    r"""Returns the class based on class name.
+def get_class_nc(
+    full_class_name: str, module_paths: Optional[List[str]] = None
+):
+    r"""Returns the class based on class name, not cached.
```
Comment (hunterhector) [inline forte/utils/utils.py:85]: please explain what is "not cached" to make this documentation clear.

File: forte/utils/utils.py [hunk 2]
```diff
@@ -127,6 +130,57 @@ def get_class(full_class_name: str, module_paths: Optional[List[str]] = None):
     return class_
 
 
+@lru_cache()
+def cached_locate(name_to_locate_class):
+    return locate(name_to_locate_class)
+
+
+def get_class(full_class_name: str, module_paths: Optional[List[str]] = None):
+    r"""Returns the class based on class name, with cache to improve speed.
```
Comment (hunterhector) [inline forte/utils/utils.py:139]: explain what is being cached (and maybe how the cache work)

File: forte/utils/utils.py [hunk 3]
```diff
@@ -127,6 +130,57 @@ def get_class(full_class_name: str, module_paths: Optional[List[str]] = None):
...
+
+    if class_ is None:
+        if module_paths:
+            raise ValueError(
+                f"Class not found in {module_paths}: {full_class_name}"
+            )
+        else:
+            raise ValueError(f"Class not found in {full_class_name}")
+
+    return class_
```
Comment (hunterhector) [inline forte/utils/utils.py:181]: I notice both methods have too many lines duplicated. The lines that are different are basically "locate" or "cached_locate". We should not write it this way. Consider merging the duplicated code as much as possible.

File: forte/utils/utils.py (hunk 1 above)
Comment (J007X) [reply forte/utils/utils.py]: This is now fixed in the new fe-factored code that combine the 2 methods into one with optional parameters

File: forte/utils/utils.py (hunk 2 above)
Comment (J007X) [reply forte/utils/utils.py]: This is now fixed: refactored code to combine the 2 methods into one ("get_class") with "cached_lookup" optional parameter.

File: forte/utils/utils.py (hunk 3 above)
Comment (J007X) [reply forte/utils/utils.py]: This is now fixed: Refactored the code to combine the 2 methods into one "get_class" with "cached_lookup" optional parameter.

**Review by hunterhector — APPROVED** [review]

**Pull request merged**